### PR TITLE
Add and update endpoints for use with django

### PR DESCRIPTION
**Pull request by ruairif**
Add schedule endpoint
Add copy endpoint
Remove all `manager` usage (reset, create, download)
Disable update and delete for projects.
Add option to initialize version_control storage at branch or commit
Add force save option in ORM

### CODE REVIEW

File: portia_server/portia_api/resources/projects.py [hunk 1]
```diff
@@ -1,29 +1,41 @@
...
-        # return ProjectDownloadResponse(
-        #     project_id, spider_ids, fmt, project_manager)
-        return Response(b'', status=HTTP_200_OK)
+        version = self.query.get('version', None)
+        branch = self.query.get('branch', None)
+        spider_id = self.kwargs.get('spider_id', None)
+        spiders = [spider_id] if spider_id is not None else None
+        if self.storage.version_control and (version or branch):
+            storage = self.storage.__class__(
+                self.storage.name, self.storage.author, version, branch)
```
Comment (sagelliv) [inline portia_server/portia_api/resources/projects.py]: We can make this a constructor method as an instance method which receives a version and branch.

File: portia_server/portia_api/resources/projects.py [hunk 2]
```diff
@@ -98,39 +109,107 @@ def status(self, *args, **kwargs):
...
-        # manager.pm.discard_changes(project_id)
+        if not self.storage.version_control and hasattr(self.storage, 'repo'):
+            raise JsonApiFeatureNotAvailableError()
+        branch = self.storage.branch
+        master = self.storage.repo.refs['refs/heads/master']
+        self.storage.repo.refs['refs/heads/%s' % branch] = master
         return self.retrieve()
 
+    @detail_route(methods=['post'])
+    def copy(self, *args, **kwargs):
```
Comment (sagelliv) [inline portia_server/portia_api/resources/projects.py]: Can we break up this function? Something like `validation`, `error handling`, `copying logic`, and the returned data?

File: portia_server/portia_api/resources/spiders.py [hunk 3]
```diff
@@ -12,3 +23,32 @@ def get_instance(self):
...
+            'project': self.project.id,
+            'spider': self.spider.id
+        }
+        if self.storage.version_control:
+            branch = self.query.get('branch', None)
+            commit = self.query.get('commit_id', None)
+            if not branch and self.storage.repo.has_branch(self.user):
+                branch = self.user
+            storage = self.storage.__class__(
+                self.storage.name, self.storage.author, commit, branch)
```
Comment (sagelliv) [inline portia_server/portia_api/resources/spiders.py]: Use a storage constructor like above.

File: portia_server/portia_api/utils/download.py [hunk 4]
```diff
@@ -173,17 +169,16 @@ def _paths(self, spiders):
...
         if isinstance(spiders, six.string_types):
             spiders = [spiders]
         spider_paths = set('spiders/%s.json' % spider for spider in spiders)
         all_files = self.list_files()
         template_paths = self._template_paths(spiders, all_files)
-        if self.version > (0, 9):
-            templates = set(itertools.chain(*template_paths.values()))
-            spider_paths = spider_paths | templates
+        templates = set(itertools.chain(*template_paths.values()))
+        spider_paths = spider_paths | templates
```
Comment (sagelliv) [inline portia_server/portia_api/utils/download.py]: Is this a bitwise OR?

**Comment by sagelliv** [comment]
:+1: After minor comments

File: portia_server/portia_api/utils/download.py (hunk 4 above)
Comment (ruairif) [inline portia_server/portia_api/utils/download.py]: Yes
```python
a = {1,2,3,4}
b = {3,4,5,6}
a | b == a.union(b)
```
Do you want me to change it to `spider_paths.union(templates)`?

File: portia_server/portia_api/utils/copy.py [hunk 5]
```diff
@@ -1,270 +1,69 @@
...
+                    for annotation in item.annotations:
+                        for extractor in annotation.extractors:
+                            extractor = extractor.with_storage(self.storage)
+                            self.project.extractors.add(extractor)
+                sample = sample.with_storage(self.storage)
+                sample.spider = new_spider
+                models.append(sample)
+            models.append(new_spider)
+        for schema in instances.get('schemas', []):
+            self.project.schemas.add(schema).save()
```
Comment (sagelliv) [inline portia_server/portia_api/utils/copy.py]: Can you do the `save` call in `self.save`?

File: portia_server/portia_api/utils/copy.py [hunk 6]
```diff
@@ -1,270 +1,69 @@
...
+        return models
+
+    def save(self, models):
+        returned_models = []
+        project = self.project
+        for model in chain(project.schemas, project.extractors, models):
+            print(model)
```
Comment (sagelliv) [inline portia_server/portia_api/utils/copy.py]: print ^

File: portia_server/portia_api/utils/copy.py [hunk 7]
```diff
@@ -1,270 +1,69 @@
...
+            model.save()
+        self.storage.commit()
+        return returned_models
```
Comment (sagelliv) [inline portia_server/portia_api/utils/copy.py]: Isn't `returned_models` always empty?

**Comment by sagelliv** [comment]
:+1:

**Pull request merged**